[INTEL MKL] Added support for quantized depthwise conv2D per channel.

bhavani-subramanian opened the pull request.

bhavani-subramanian committed:
Enabled support for quantized depthwise conv2D per channel.

bhavani-subramanian commented:
@penpornk Can you please take a look at this PR? Thank you.

reedwm commented:
I'm not familiar with this code, so I'll unassign myself and leave the review to @penpornk

penpornk commented on tensorflow/core/ops/mkl_nn_ops.cc:
```diff
@@ -804,6 +803,204 @@ REGISTER_OP("_MklDepthwiseConv2dNativeBackpropFilter")
+      ShapeHandle unused;
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 1, &unused));
+      for (int i = 3; i < 9; ++i) {
+        TF_RETURN_IF_ERROR(c->WithRank(c->input(i), 0, &unused));
```
Are `min_filter` and `max_filter` here always scalars (and not vectors)?

bhavani-subramanian replied:
Good catch! They can be vectors. I have modified code to reflect this change.

penpornk commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2884,6 +2884,7 @@ REGISTER_OP("QuantizedConv2DWithBiasSignedSumAndReluAndRequantize")
       TF_RETURN_IF_ERROR(c->WithRank(c->input(4), 0, &unused));
       TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(5), 1, &channel));
       TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(6), 1, &channel));
+      // TODO(bhavanis): Need to use c->WithRankAtMost(...) for 7th input?
```
Should the shapes of `min_output` and `max_output` here be set to `channel`?

bhavani-subramanian replied:
They are scalar since activations are not requantized per channel. I have added this explanation as a comment in the code.

penpornk replied:
Sounds great. Thank you!

penpornk reviewed (changes requested):
I'm so sorry for my delay! 
Do you think you can make all these ops one fused op, i.e., `_MklQuantizedDepthwiseConv2D` (or `_MklFusedQuantizedDepthwiseConv2D`) with an attribute storing a list of all fused ops, like `_FusedConv2D`?
https://github.com/tensorflow/tensorflow/blob/123e0d56cd981cad847c120c7339b807ee7cbf64/tensorflow/core/ops/nn_ops.cc#L322-L342

If you'd rather keep them like this for now (to make sure this PR is done in time for the next TF release) and change it later, I'm fine with that too (if the API folks are okay with it).

bhavani-subramanian committed:
Addressed review comments.

bhavani-subramanian committed:
Ran Clang format checks.

bhavani-subramanian commented on tensorflow/core/api_def/base_api/api_def_QuantizedDepthwiseConv2D.pbtxt:
```diff
@@ -0,0 +1,4 @@
+op {
+  graph_op_name: "QuantizedDepthwiseConv2D"
```
Done.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -379,7 +379,14 @@ class MklConvFwdPrimitiveFactory : public MklPrimitiveFactory<T> {
       } else if (post_op_param.name == "sum") {
         DCHECK_EQ(post_op_param.param.size(), 1);
-      } else if (post_op_param.name != "output_scale") {
+        key_creator.AddAsKey(post_op_param.name);
```
Agree. Removed code that added `param` more than once.

bhavani-subramanian commented on tensorflow/core/kernels/mkl_conv_ops.cc:
```diff
@@ -2073,22 +2077,25 @@ REGISTER_KERNEL_BUILDER(Name("QuantizedConv2DWithBiasSumAndReluAndRequantize")
-// Register a templatized implementation of MklQuantizedConv2DWithBiasAndRelu.
+
+// Register a templatized implementation of
+// MklQuantizedConv2DWithBiasSumAndRelu.
```
Clang format prompted me to break it up into two lines. Do you want me to revert it to the original code?

penpornk replied:
Thank you!

bhavani-subramanian commented on tensorflow/core/kernels/mkl_quantized_conv_ops_test.cc:
```diff
@@ -454,5 +454,291 @@ TEST_F(QuantizedConv2DTest, OddPaddingBatch) {
+                   .Finalize(node_def()));
+  TF_ASSERT_OK(InitOp());
+
+  const int depth = 2;
```
Done.

bhavani-subramanian commented on tensorflow/core/ops/mkl_nn_ops.cc:
```diff
@@ -804,6 +803,204 @@ REGISTER_OP("_MklDepthwiseConv2dNativeBackpropFilter")
+    .Output("mkl_max_output: uint8")
+    .Attr("Tinput: quantizedtype")
+    .Attr("Tfilter: quantizedtype")
+    // Additional attribute "T" for enabling MKL to TF conversion
```
Done.

bhavani-subramanian commented on tensorflow/core/ops/mkl_nn_ops.cc:
```diff
@@ -804,6 +803,204 @@ REGISTER_OP("_MklDepthwiseConv2dNativeBackpropFilter")
+      ShapeHandle unused, channel;
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(2), 0, &unused));
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(3), 0, &unused));
+      TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(4), 1, &channel));
```
Agree. Changed it to `unused`.

bhavani-subramanian commented:
@penpornk Except for combining all quantized depthwise ops into one fused op (which I'll be changing later), I think I have addressed all your review comments. Thanks

bhavani-subramanian committed:
Reverted change made to post-ops key generation.

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/quantized-depthwise-conv-per-channel

bhavani-subramanian committed:
Update mkl_layout_pass_test.cc

bhavani-subramanian committed:
Update mkl_layout_pass_test.cc

penpornk reviewed (commented):
Thank you for all the fixes! I think things are fine now and the only change left is the fused op.

bhavani-subramanian commented:
Thanks for the review, @penpornk! Is it okay if we implement the fused op later (after the next TF release) so that this PR gets done now?

penpornk commented:
@bhavani-subramanian I'm okay with that (unless API folks don't like it). I'll tag API review then.

penpornk reviewed (changes requested):
Talking about API, I believe you'll have to update the API golden files. Please run 
```
bazel build //tensorflow/tools/api/tests/api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True --update_goldens=True
```
And add the changed files to the PR. There should be changes for
`tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt` and 
`tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt`

bhavani-subramanian committed:
Updated API golden files.

penpornk reviewed (approved):
Thank you! Approving for non-API-change changes.

bhavani-subramanian commented:
@penpornk I have updated the API golden files.

alextp commented:
Adding ready to pull to trigger internal changelist creation so this shows up in the tf api review process.

mdfaijul commented on tensorflow/core/ops/mkl_nn_ops.cc:
```diff
@@ -804,6 +803,208 @@ REGISTER_OP("_MklDepthwiseConv2dNativeBackpropFilter")
+REGISTER_OP("_MklQuantizedDepthwiseConv2DWithBiasAndReluAndRequantize")
+    .Input("input: Tinput")
+    .Input("filter: Tfilter")
+    .Input("bias: float")
```
@bhavani-subramanian need to change `.Input("bias: float")` -> .Input("bias: Tbias")`

penpornk replied:
The API review is happening this afternoon so we are pulling to create an internal changelist now. This change is not related to API and can be added to the changelist later (through this PR). Or I can make the change directly as a separate commit as well.

bhavani-subramanian replied:
@penpornk Could you please make this change as a separate commit? Thanks.

penpornk replied:
Sure, I'll do that after this PR is merged. (If it gets through the API review.)

bhavani-subramanian replied:
Thanks @penpornk!

penpornk replied:
Change submitted in https://github.com/tensorflow/tensorflow/commit/978dba096289fd03b9baa38c82e966eba03c6e1a.

penpornk commented:
@rthadur Friendly ping for pulling.
